Filter out ANSI escape codes from logs when retrieving logs from the dashboard

kenmcheng opened the pull request:
<!-- Thank you for your contribution! Please review https://github.com/ray-project/ray/blob/master/CONTRIBUTING.rst before opening a pull request. -->

<!-- Please add a reviewer to the assignee section when you create a PR. If you don't have the access to it, we will shortly find a reviewer and assign them to your PR. -->

## Why are these changes needed?
Based on issue #52346, log messages may contain ANSI escape codes, such as `[1;31m`, when highlighted with colors in the terminal.
This PR is intended to filter out ANSI escape codes from logs when retrieving them for use in the dashboard.

![Screenshot 2025-05-21 212327](https://github.com/user-attachments/assets/c1004495-72c6-4d22-b521-aee6bb91e48e)

The same message in the terminal is still colored
![Screenshot 2025-05-21 212253](https://github.com/user-attachments/assets/c56f016a-acd0-4193-af00-619c913076bb)

<!-- Please give a short summary of the change and the problem this solves. -->

## Related issue number
Closes #52346
<!-- For example: "Closes #1234" -->

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
    - [ ] I've added any new APIs to the API Reference. For example, if I added a
           method in Tune, I've added it in `doc/source/tune/api/` under the
           corresponding `.rst` file.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

<!-- CURSOR_SUMMARY -->
---

> [!NOTE]
> Strips ANSI escape codes from dashboard log streaming and adds tests verifying colored output is filtered.
> 
> - **Dashboard backend**:
>   - Strip ANSI escape codes from streamed logs in `python/ray/dashboard/modules/state/state_head.py` by introducing `ANSI_ESC_PATTERN` and applying it to the first and subsequent log chunks in `get_logs`.
> - **Tests**:
>   - Add assertions in `python/ray/tests/test_state_api_log.py` to verify colored log messages are returned without ANSI codes while preserving text content.
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 481494fa9553afed31ffc26f7f5dadb6484c9a16. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>
<!-- /CURSOR_SUMMARY -->

kenmcheng committed:
Filter out ANSI escape codes from logs when retrieving logs from the dashboard

When the log content is highlighted with colors in the terminal, the text contains ANSI escape codes such as '[1;31m', which will otherwise surface on non-terminal interfaces.
In order to mitigate the disturbance, this commit attempts to filter out escape codes.

Signed-off-by: kenmcheng <kengmingcheng@outlook.com>

kenmcheng commented:
Thanks for relying. Take your time and free feel to ping me when you have time to look back.

kenmcheng commented:
Commenting to keep this PR  from becoming stale. I can also submit a new one once the reviewers are available.

alanwguo commented on python/ray/dashboard/modules/state/state_head.py:
```diff
@@ -246,11 +247,15 @@ async def get_actor_fn(actor_id: ActorID) -> Optional[ActorTableData]:
 
         logs_gen = self._log_api.stream_logs(options, get_actor_fn)
+        # For filtering ANSI escape codes; the byte string used in the regex is equivalent to r'\x1b\[[\d;]+m'.
+        ansi_esc_pattern = re.compile(b"\x1b\\x5b[(\x30-\x39)\x3b]+\x6d")
```
let's compile this once and share it across function calls.

alanwguo reviewed (commented):
thanks for the contribution!

can you add a basic unit test for this?

kenmcheng committed:
Merge branch 'master' into 52346-ansi-esc-code

kenmcheng committed:
1. let ANSI_ESC_PATTERN be only compiled once
2. Add an integration test for checking ANSI escape codes are filtered out correctly

Signed-off-by: kenmcheng <kengmingcheng@outlook.com>

kenmcheng commented on python/ray/dashboard/modules/state/state_head.py:
```diff
@@ -48,6 +49,9 @@
 )
 
+# For filtering ANSI escape codes; the byte string used in the regex is equivalent to r'\x1b\[[\d;]+m'.
+ANSI_ESC_PATTERN = re.compile(b"\x1b\\x5b[(\x30-\x39)\x3b]+\x6d")
```
This suggestion causes error during compiling the pattern
```
print-byte.py:18: FutureWarning: Possible nested set at position 2
  ptn = re.compile(b'\x1b\x5b[(\x30-\x39)\x3b]+\x6d')
```

kenmcheng committed:
1. Add a path arugment as a flag for flitering ANSI escape codes
2. Adjuest corresponding test case

Signed-off-by: kenmcheng <kengmingcheng@outlook.com>

kenmcheng committed:
Merge branch 'master' into 52346-ansi-esc-code

kenmcheng commented on python/ray/dashboard/modules/state/state_head.py:
```diff
@@ -249,8 +253,10 @@ async def get_actor_fn(actor_id: ActorID) -> Optional[ActorTableData]:
         try:
             first_chunk = await logs_gen.__anext__()
+            # Filter ANSI escape codes
+            filtered_first_chunk = ANSI_ESC_PATTERN.sub(b"", first_chunk)
```
Added. Should I also add a corresponding argument on clients (e.g. dashboard web UI)?

kenmcheng commented on python/ray/tests/test_state_api_log.py:
```diff
@@ -1284,6 +1284,37 @@ def verify():
+        lines = get_log(actor_id=actor_id, suffix="out")
+        joined_lines = "".join(lines)
+        assert NO_COLOR_MESSAGE in joined_lines
+        assert UNCOLORED_MESSAGE in joined_lines
```
Done. I also added another assertion for the requset without filtering.

alanwguo commented on python/ray/dashboard/modules/state/state_head.py:
```diff
@@ -225,6 +229,7 @@ async def get_logs(self, req: aiohttp.web.Request) -> aiohttp.web.Response:
             suffix=req.query.get("suffix", "out"),
             attempt_number=req.query.get("attempt_number", 0),
         )
+        filtering_ansi_esc_code = req.query.get("filter_ansi_code", False)
```
let's update this to check for the value is == "true"

alanwguo commented on python/ray/util/state/api.py:
```diff
@@ -1184,6 +1184,7 @@ def get_log(
     submission_id: Optional[str] = None,
     attempt_number: int = 0,
     _interval: Optional[float] = None,
+    filter_ansi_code: bool = None,
```
please update default to False instead of None

alanwguo reviewed (commented):
Yes, in the ray dashboard, please set the value to true so the UI will filter the ansi codes.

kenmcheng committed:
1. Add a check for 'filter_ansi_code' == true
2. Add a corresponding test case for 'filter_ansi_code' = "random value"
3. Add the 'filter_ansi_code' query variable to the dashboard web client
4. Update related test cases for the dashboard web client

Signed-off-by: kenmcheng <kengmingcheng@outlook.com>

alanwguo reviewed (approved):
Thanks for the fix! LGTM except for linter error

kenmcheng committed:
Add a trailing comma to fix the linter error

Signed-off-by: kenmcheng <kengmingcheng@outlook.com>

Commit:
Merge branch 'master' into 52346-ansi-esc-code

kenmcheng commented:
The new failure isn't likely to be the same as the previous failure. There is another PR #57004  failing on the same [case](https://buildkite.com/ray-project/premerge/builds/51633/steps/canvas).
I am not sure if this has been fixed on the latest main branch.

jjyao committed:
Merge branch 'master' into 52346-ansi-esc-code
